### PR TITLE
Add basic functional tests to mintlayer

### PR DESCRIPTION
**Pull request by TheQuantumPhysicist**
Inherit Bitcoin's way of writing functional tests, where this system launches nodes, connects them together, and can be then controlled using RPC.

### CODE REVIEW

**Comment by altonen** [comment]
Should this be merged to staging first and then to master? Your commits are missing signatures which in the past have prevented merging

**Comment by TheQuantumPhysicist** [comment]
> Should this be merged to staging first and then to master? Your commits are missing signatures which in the past have prevented merging

Yes, of course. These should go to staging.

I never used signatures on github. Let's see if we're gonna have a problem because of that.

File: node/build.rs [hunk 1]
```diff
@@ -1,7 +1,62 @@
...
 
+use std::env;
+use std::ffi::OsStr;
+use std::fs;
+use std::path::Path;
+
+fn write_functional_tests_config_file()
+{
+    let cargo_manifest_dir = env!("CARGO_MANIFEST_DIR");
+    let profile = env::var("PROFILE").unwrap();
```
Comment (altonen) [inline node/build.rs:11]: unwrap

File: test/functional/custom-types.json [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            ]
+        },
+        "Destination": {
+            "type": "enum",
+            "type_mapping": [
+                [ "Pubkey", "H256" ],
+                [ "CreatePP", "DestinationCreatePP" ],
+                [ "CallPP", "DestinationCallPP" ],
+                [ "ScriptHash", "H256" ],
+                [ "PubkeyHash", "Vec<u8>" ]
```
Comment (altonen) [inline test/functional/custom-types.json:25]: We don't have `Destionation::PubkeyHash` anymore. I think this file should be checked against `utxo-tester`'s `custom-types.json` as right now it's known to be up to date.

File: test/functional/example_test.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                    destination=utxo.DestPubkey(alice.public_key)
+                ),
+                utxo.Output(
+                    value=100,
+                    header=0,
+                    destination=utxo.DestPubkey(alice.public_key)
+                ),
+
+            ]
+        ).sign(alice)
```
Comment (altonen) [inline test/functional/example_test.py:124]: I think this is not correct, refer to the `utxo-tester` repo

File: test/functional/feature_alice_bob_test.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                utxo.Input(utxos[0]),
+            ],
+            outputs=[
+                utxo.Output(
+                    value=50,
+                    header=0,
+                    destination=utxo.DestPubkey(bob.public_key)
+                ),
+            ]
+        ).sign(alice)
```
Comment (altonen) [inline test/functional/feature_alice_bob_test.py:79]: Same here with the signature, it was updated recently

File: test/functional/test_framework/mintlayer/utxo.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
+import substrateinterface
```
Comment (altonen) [inline test/functional/test_framework/mintlayer/utxo.py:1]: This file is also old. Is the plan to just merge this to staging and updated it later to be up to date with `utxo-tester`?

**Review by altonen — COMMENTED** [review]
I think the code is not fully up to date with `utxo-tester` and the general guideline to my understanding is to sign commits. I think it's in the employee manual but let's see what Github says

File: test/functional/feature_alice_bob_test.py (hunk 4 above)
Comment (TheQuantumPhysicist) [reply test/functional/feature_alice_bob_test.py:79]: Probably I'm using the "old" version of utxo tester... again... another motivation to merge this ASAP. I can't keep up. It's at least two other people working on this in parallel with me.

**Review by altonen — APPROVED** [review]

**Pull request merged**